ord_schema/frozen_message.py

```diff
@@ -0,0 +1,111 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Wrappers and utilities for handling protocol buffers in Python."""
+
+import collections
+import dataclasses
+from typing import Union
+
+import google.protobuf.message
+
+_MESSAGE_TYPES = (
+    collections.abc.MutableMapping,  # Proto map.
+    google.protobuf.message.Message,  # Generic submessage.
+)
+
+
+@dataclasses.dataclass(eq=True, frozen=True)
+class FrozenMessage(collections.abc.Mapping):
+    """Container for a protocol buffer that does not allow edits.
+
+    Notes:
+        * For standard scalar values, it is not possible to distinguish between
+          default values and explicitly set values that match the default. If
+          the default is a valid value, add the `optional` label to the field.
+          See https://github.com/Open-Reaction-Database/ord-schema/pull/174.
+        * For `optional` scalar values and all submessage fields, exceptions
+          are raised if the user attempts to access an undefined attribute
+          (AttributeError), access an undefined map key (KeyError), or set
+          any attribute or map value (dataclasses.FrozenInstanceError).
+        * I considered adding a `raise_on_error` option that would return None
+          instead of raising AttributeError or KeyError when requesting unset
+          values. However, this breaks the guarantee that `hasattr` returns
+          False for unset `optional` scalar values and submessages.
+    """
+    _message: Union[_MESSAGE_TYPES]
+
+    def __getattr__(self, name):
+        """Fetches a message attribute, if it exists.
+
+        Notes:
+            * If `name` is a submessage, a new FrozenMessage wrapping the
+              submessage is returned.
+            * If `name` has not been set explicitly, an AttributeError is
+              raised. This allows for distinguishing between default values and
+              values that were set explicitly (to the default).
+
+        Args:
+            name: Text attribute name.
+
+        Returns:
+            The requested attribute. If the attribute is a submessage, a new
+            FrozenMessage is returned.
+
+        Raises:
+            AttributeError: if `name` has not been set explicitly.
+        """
+        try:
+            if not self._message.HasField(name):
+                raise AttributeError(f'attribute "{name}" has not been set')
+        except ValueError:
+            pass  # The request attribute is not a submessage.
+        value = getattr(self._message, name)
+        if isinstance(value, _MESSAGE_TYPES):
+            return FrozenMessage(value)
+        if hasattr(value, 'append'):
+            return tuple(value)  # Make repeated fields immutable.
```

Comment:
Does this need to recurse and return `tuple(map(lambda v: FrozenMessage(v) if isinstance(v, _MESSAGE_TYPES) else v, values))` or `tuple(FrozenMessage(v) if isinstance(v, _MESSAGE_TYPES) else v for v in values)`?

Reply:
```
>>> message = reaction_pb2.Reaction()
>>> message.workup.add(type=2)
type: ADDITION

>>> frozen = frozen_message.FrozenMessage(message)
>>> frozen
FrozenMessage(_message=workup {
  type: ADDITION
}
)
>>> frozen.workup[0].type = 3
>>> frozen
FrozenMessage(_message=workup {
  type: TEMPERATURE
}
)
```

Reply:
Good catch; I'll revise.